components/data-feed-service/service/data-feed-poll-task.go

```diff
@@ -0,0 +1,209 @@
+	manager *cereal.Manager
+}
+
+type DataFeedPollTaskParams struct {
+	AssetPageSize   int32
+	ReportsPageSize int32
+	FeedInterval    time.Duration
+	NextFeedStart   time.Time
+	NextFeedEnd     time.Time
+}
+
+type DataFeedPollTaskResults struct {
+	NodeIDs   map[string]NodeIDs
+	FeedStart time.Time
+	FeedEnd   time.Time
+}
+
+// ClientID the node ID assigned by chef client
+// ComplianceID the node ID in a compliance report
+type NodeIDs struct {
+	ClientID     string
+	ComplianceID string
+}
+
+func NewDataFeedPollTask(dataFeedConfig *config.DataFeedConfig, connFactory *secureconn.Factory, db *dao.DB, manager *cereal.Manager) (*DataFeedPollTask, error) {
+
+	cfgMgmtConn, err := connFactory.Dial("config-mgmt-service", dataFeedConfig.CfgmgmtConfig.Target)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not connect to config-mgmt-service")
+	}
+
+	return &DataFeedPollTask{
+		cfgMgmt: cfgmgmt.NewCfgMgmtClient(cfgMgmtConn),
+		db:      db,
+		manager: manager,
+	}, nil
+}
+
+func (d *DataFeedPollTask) Run(ctx context.Context, task cereal.Task) (interface{}, error) {
+	params := DataFeedWorkflowParams{}
+	err := task.GetParameters(&params)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse task parameters")
+	}
+	log.Debugf("DataFeedPollTask.Run params %v", params)
+	log.Debugf("DataFeedPollTask.Run FeedStart %v", params.FeedStart)
+	log.Debugf("DataFeedPollTask.Run FeedEnd %v", params.FeedEnd)
+	log.Debugf("DataFeedPollTask.Run NextStart %v", params.PollTaskParams.NextFeedStart)
+	log.Debugf("DataFeedPollTask.Run NextEnd %v", params.PollTaskParams.NextFeedEnd)
+	now := time.Now()
+	params = d.getFeedTimes(params, now)
+	feedStartTime := params.PollTaskParams.NextFeedStart
+	feedEndTime := params.PollTaskParams.NextFeedEnd
+	destinations, err := d.db.ListDBDestinations()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get destinations from db")
+	}
+	params.FeedStart = feedStartTime
+	params.FeedEnd = feedEndTime
+	params.PollTaskParams.NextFeedStart = feedEndTime
+	params.PollTaskParams.NextFeedEnd = feedEndTime.Add(params.PollTaskParams.FeedInterval)
+
+	if len(destinations) == 0 {
+		log.Info("DataFeedPollTask.Run no destinations returning")
+		return nil, nil
+	}
+
+	nodeIDs, err := d.GetChangedNodes(ctx, params.PollTaskParams.AssetPageSize, feedStartTime, feedEndTime)
+	log.Debugf("DataFeedPollTask nodes %v", nodeIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	params.NodeIDs = nodeIDs
+
+	log.Debugf("Updated next feed interval start, end: %s, %s", params.PollTaskParams.NextFeedStart.Format("15:04:05"), params.PollTaskParams.NextFeedEnd.Format("15:04:05"))
+	log.Debugf("Updated feed interval start, end: %s, %s", params.FeedStart.Format("15:04:05"), params.FeedEnd.Format("15:04:05"))
+
+	// we must update the workflow params to ensure the nest schedule has the update interval times
+	err = d.manager.UpdateWorkflowScheduleByName(context.Background(), dataFeedScheduleName, dataFeedWorkflowName,
```

Comment:
Here you can use the context that we were passed as an argument rather than context.Background().

Reply:
I'm wondering if this is the right place to update the schedule.  At this point we don't know if the work has actually been done yet. Should we wait to update the schedule until the data is actually sent to the destinations so that intervals are retried if they fail?

Reply:
At the moment yes. Current implementation is fail safely and  move on. Retries aren't trivial, we could get stuck on an interval with corrupt data, or fall behind and then the lag correction would skip intervals, we could try and add multiple intervals to the poll task so that retries are included with the current interval but then we could end up constantly retrying multiple intervals. We could end up retrying an interval that's data has already been superseded by a subsequent interval. I went for a keep it simple approach. Errors and retries are a roadmap feature! I'm actually going to do it slightly earlier so that the schedule is updated on error and if there is no work to do.